apps/webpack.config.js

```diff
@@ -601,6 +603,15 @@ function createWebpackConfig({
           }
           return file;
         },
+        ...(process.env.DEV
```

Comment:
Thinking about how we're using `CDO.optimize_webpack_assets` on the dashboard side of things, would it make sense to use the same CDO flag here? (Trying to remember what the webpack.config has access to.)

Reply:
@bencodeorg ^ Referring to your new PR https://github.com/code-dot-org/code-dot-org/pull/55925/files

Reply:
Yeah I was wondering about that...I was thinking that wouldn't work in a non-Rails context (and it doesn't appear we currently can access that sort of config easily in `apps` code)? I could be totally wrong though. We could add another flag to `start`, `build`, etc like `OPTIMIZE_WEBPACK_ASSETS=1` if someone is doing a string search at least? Or that might be more confusing 😬

Reply:
I personally wouldn't want to use/reference CDO in JS build bits. As @bencodeorg mentions, we'd have to build the bridge, but moreover I don't think its a bridge that should be built.

Reply:
Crazy idea, possibly bad.... what if we sniff if the file exists from the rails side too, instead of indexing off a flag (`CDO.optimize_webpack_assets`)?

```
- if file_exists('js/code-studio-common.js')
      %script{src: webpack_asset_path('js/vendors.js')}
      %script{src: webpack_asset_path('js/code-studio-common.js')}
```

It /probably/ doesn't matter if we add an extra fs syscall or two to the default page template, but it'd also be really easy to memoize file_exists in production mode (in dev mode you want it to always check so you don't have to restart `dashboard-server` if you change JS build type).

Reply:
Possibly the right logic would be `if file_exists('js/code-studio-common.js') || CDO.optimize_webpack_assets`, which would give you 404s if you had done the wrong JS build type for current prod-like environment.

Reply:
Are you thinking it's a performance or just general code improvement to do this by checking for existence of the custom webpack chunk files? I was wondering if we have any existing usages of `CDO.optimize_webpack_assets` that gets hit as part of our production code, looks like any time `webpack_asset_path` is called that config value is checked:

https://github.com/code-dot-org/code-dot-org/blob/59fc56e18376beeda7df10952b7b456604544355/lib/cdo/asset_helper.rb#L35-L43

I think I find what I have here a bit clearer than checking whether one of those webpack files exists, but open to changing if if you're feeling like it's a better direction.

Reply:
My thinking is, that I'm not personally confident that DEV(js)==CDO.optimize_webpack_assets(rails) in all conditions. It could well be true, but it requires a lot of knowledge to reason about that I don't have. On the other hand, by looking at the filesystem (shared between webpack node and rails), we know that both systems are indexing off the same state, so you can't have a state tear.

I'm not worried about the performance of `CDO.optimize_webpack_assets`, I'm worried that it might not have a one-to-one correspondence with DEV=1 being set during the yarn build.

I /was/ slightly worried that file_exists on every template render could conceivably slow things down, but rails is probably doing so many filesystem calls that its lost in the noise (and I suggested an optimization for that). But the optimization isn't relevant unless we thought file_exists was a more consistent/better check to match DEV than the CDO flag.

Reply:
It seems totally reasonable to me to set DEV=1 if (and only if) `!CDO.optimize_webpack_assets`, and we could do it via similar mechanism to `generateSharedConstants.rb`. However I am also not seeing a need for this -- 99% of the time the developer wants to run with DEV=1, and if they want to run with optimized JS they will be consulting special instructions and it will be easy enough to set optimized js there "manually" in that case.

or am i missing a new, increased need for running optimized js locally?

Reply:
I'm unaware of a new need to run optimized JS locally more regularly :). I think mostly it's that we (especially me) are mucking around in code we don't fully understand, so wanted to be cognizant of any changes we're

I feel more confident having your input that this seems reasonable though Dave, as someone who is very familiar with our JS build process.

I'm kind of inclined to get this merged (including the two branch off branch PRs) and out to folks and see if we run into any issues -- thoughts?

Reply:
Yeah, I felt the same as Ben, we know how much we don't know! I think Dave weighing in here is the kicker, and it makes sense to merge, get it into folk's hands, and handle any future changes as a followup (if anything does crop up).